### PR TITLE
Add analyzer to suggest top level route registration

### LINKED ISSUE #35759: Fixer/Analyzer: Remove UseEndpoints() and hoist route registrations top level to WebApplication
With the new hosting model, it's possible to register routes top level instead of using UseEndpoints. To help customer learn about this, we can have an fixer that can hoist the route registrations to hang off WebApplicatin:

```C#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.UseRouting();
app.UseEndpoints(endpoints =>
{
    endpoints.MapGet("/", () => "Hello World!");
});
```

We would suggest this:

```C#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.UseRouting();
app.MapGet("/", () => "Hello World!");

app.Run();
```

Stretch goal is to also remove UseRouting but we would need to only remove it if it wasn't intentionally put top level (like set to re-run middleware).

### PR DESCRIPTION
**Pull request by pedro-camargo11**
# ASP0014 analyzer: Use top level route registration.
Add analyzer ASP0014 to WAB analyzers. Suggests top level route registration instead of using UseEndpoints.

Fires in the following scenario:
```
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.UseRouting();
app.UseEndpoints(endpoints =>
{
    endpoints.MapGet("/", () => "Hello World!");
});
```

Fixes #35759

### CODE REVIEW

File: src/Framework/AspNetCoreAnalyzers/test/WebApplicationBuilder/UseTopLevelRouteRegistrationInsteadOfUseEndpointsTest.cs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    {
+        //arrange
+        var source = TestSource.Read(@"
+using Microsoft.AspNetCore.Builder;
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+app.UseRouting();
+app./*MM*/UseEndpoints(endpoints =>
+{
+    endpoints.MapGet(""/"", () => ""Hello World!"");
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/test/WebApplicationBuilder/UseTopLevelRouteRegistrationInsteadOfUseEndpointsTest.cs:72]: Add a test for other map methods, e.g. MapPost

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/DiagnosticDescriptors.cs [hunk 2]
```diff
@@ -106,4 +106,13 @@ internal static class DiagnosticDescriptors
         DiagnosticSeverity.Warning,
         isEnabledByDefault: true,
         helpLinkUri: "https://aka.ms/aspnet/analyzers");
+
+    internal static readonly DiagnosticDescriptor UseTopLevelRouteRegistrationsInsteadOfUseEndpoints = new(
+        "ASP0014",
+        "Suggest using app.MapGet over of using app.UseEndpoints",
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/DiagnosticDescriptors.cs:112]: This message should be more generic to cover other Map methods, e.g. MapPost, MapPut, etc.

**Review by captainsafia — APPROVED** [review]

**Comment by erichiller** [comment]
Could you add some docs for this, I didn't get what it was wanting me to do, finally I figured it out. Maybe docs like this:

***

**ASP0014**

x        | Description
--------|------
Code    | `ASP0014`
Title   | _Suggest using top level route registrations_
Message | _Suggest using top level route registrations instead of `UseEndpoints`_ (`UseEndpoints` will be filled in with whatever method is being used and should be replaced)
Name    | [`UseTopLevelRouteRegistrationsInsteadOfUseEndpoints`][asp-analyzer-cs]
Introduced | [PR #42937]
Source     | [`WebApplicationBuilderAnalyzer.cs`]
Tests      | [Tests][ASP0014-TEST]

Fix by replacing [`UseEndpoints`] with the same method calls directly on [`WebApplication`] which also implements [`IEndpointRouteBuilder`]

Example, old:
```cs
webApp.UseEndpoints(endpointRouteBuilder => {
    endpointRouteBuilder.MapGrpcService<MyService>();
});
```
Example, new:
```cs
webApp.MapGrpcService<MyService>();
```

Suppress via attribute:
```cs
[SuppressMessage("Usage", "ASP0014:Suggest using top level route registrations")]
```

[`IEndpointRouteBuilder`]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.iendpointroutebuilder
[`WebApplication`]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.webapplication
[`UseEndpoints`]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.endpointroutingapplicationbuilderextensions.useendpoints
[PR #42937]: https://github.com/dotnet/aspnetcore/pull/42937/
[ASP0014-TEST]: https://github.com/dotnet/aspnetcore/pull/42937/files/0c41eda09d10e70c5b9c3b5781abb61cc5980c73
[asp-analyzer-cs]: https://github.com/dotnet/aspnetcore/blob/main/src/Framework/AspNetCoreAnalyzers/src/Analyzers/DiagnosticDescriptors.cs
[`WebApplicationBuilderAnalyzer.cs`]: https://github.com/dotnet/aspnetcore/blob/main/src/Framework/AspNetCoreAnalyzers/src/Analyzers/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs





*****

**Comment by captainsafia** [comment]
@erichiller Thanks for the feedback, Eric! We can definitely add some docs for this.

I'll admit I have to dust off some of the cobwebs around how we handle docs for new analyzers. I had thought that some of this was automated but turns out we'll actually have to populate some content for the new analyzers that we added in .NET 7 over at https://docs.microsoft.com/en-us/aspnet/core/diagnostics/code-analysis?view=aspnetcore-6.0.

I've added #44009 to track adding these docs.

**Comment by HarelM** [comment]
I've seen this warning in my code and moved from endpoints to top level route.
This completely broke my app.
Can the docs/analyzers be improved so that this won't happen to other people or better explain how to change it so that it won't break the app API?
The following is the commit I did in order to revert back to a working state, I'd love to know what I did wrong in the first place and how to properly resolve the warning I get...

**Comment by captainsafia** [comment]
> I've seen this warning in my code and moved from endpoints to top level route. This completely broke my app. Can the docs/analyzers be improved so that this won't happen to other people or better explain how to change it so that it won't break the app API? The following is the commit I did in order to revert back to a working state, I'd love to know what I did wrong in the first place and how to properly resolve the warning I get...

I took a look at the app linked. I believe it might be related to the fact that you invoke `UseRouting` and `UseEndpoints` via your own ordering in the target application (see [this](https://github.com/dotnet/aspnetcore/issues/45903#issuecomment-1375925538)).

**Comment by HarelM** [comment]
Thanks for the tip @captainsafia!

**Pull request merged**